### PR TITLE
[TensorExpr] Add a pass for trimming JIT graphs.

### PR DESCRIPTION
**Pull request by ZolotukhinM**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70475
* #70375
* #70374
* __->__ #66847
* #70535
* #66515
* #66514

Trimming means that we try to remove a small portion of the graph while
keeping it valid, and we try performing this step that N times. This is
useful for debugging when we try to find a minimal example reproducing
the issue at hand.

Differential Revision: [D31751397](https://our.internmc.facebook.com/intern/diff/D31751397)

### CODE REVIEW

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Comment by eellison** [comment]
ccc @davidberard98 -  something like this may be useful for repro'ing nvfuser failures

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 1]
```diff
@@ -916,18 +927,37 @@ BufHandle TensorExprKernel::bindSymbolicShapeInput(
...
+        if (!outputs_set.count(input)) {
+          bufs_.emplace(input, bufHandle.node());
+        } else {
+          // If 'input' is also a graph output, then we need to produce a copy
+          std::vector<DimArg> inputTensorDims;
+          for (size_t i = 0; i < bufHandle.dims().size(); i++) {
+            inputTensorDims.emplace_back(
+                DimArg(bufHandle.dims()[i], "i" + c10::to_string(i)));
+          }
+          result = Compute(
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/kernel.cpp:949]: Can we try to combine the copying here with the copying below? The one below includes handling strides, but that can probably be an argument and the logic of copying would be in one place then.

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 2]
```diff
@@ -916,18 +927,37 @@ BufHandle TensorExprKernel::bindSymbolicShapeInput(
 
 Tensor TensorExprKernel::bindInput(const torch::jit::Value* input) {
   auto const& t = input->type();
+  auto const& outputs = input->owningGraph()->outputs();
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/kernel.cpp:930]: Why are these changes to `kernel.cpp` in this PR? They are not directly related to "trimming JIT graphs", right?

File: torch/csrc/jit/tensorexpr/graph_opt.h [hunk 3]
```diff
@@ -69,6 +69,16 @@ TORCH_API std::shared_ptr<Graph> removeGraphOutput(
 TORCH_API std::shared_ptr<Graph> replaceListOutputWithTuple(
     const std::shared_ptr<Graph>& graph);
 
+// Perform \p ITERS rounds of "trimming" for the given \p GRAPH.
+//
+// Trimming means that we try to remove a small portion of the graph while
```
[the remark this answers is not in the document]
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/graph_opt.h:74]: Tbh, that's exactly the contract this function provides. What exact portion is removed and how it is done is not guaranteed - that's an implementation detail and should not matter/be relied on.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 2 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/kernel.cpp:930]: That's correct, this fixes the behavior when a graph input is used directly as an output. We don't usually get such graphs, but that becomes common when we try trimming them.

I'll extract it to a separate PR.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 1 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/kernel.cpp:949]: Let me see what we can do there!

Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/kernel.cpp]: Please take a look at #70535! I extracted the changes to a separate PR as you suggested below and refactored the logic a little.

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Review by navahgar — APPROVED** [review]
LGTM

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31751397).

**Pull request closed without merging**